### PR TITLE
Support HTTP_X_FORWARDED_PROTO through multiple proxy levels

### PR DESCRIPTION
**Pull request by bpicolo**
| Q             | A
| ------------- | ---
| Is bugfix?    | yes
| New feature?  | no
| Breaks BC?    | yes/no   Depends on the behavior people might be relying on? Doesn't break any well-defined behavior.
| Tests pass?   | yes, + new test added
| Fixed issues  | #13451

### CODE REVIEW

File: framework/web/Request.php [hunk 1]
```diff
@@ -874,7 +874,7 @@ public function getQueryString()
     public function getIsSecureConnection()
     {
         return isset($_SERVER['HTTPS']) && (strcasecmp($_SERVER['HTTPS'], 'on') === 0 || $_SERVER['HTTPS'] == 1)
-            || isset($_SERVER['HTTP_X_FORWARDED_PROTO']) && strcasecmp($_SERVER['HTTP_X_FORWARDED_PROTO'], 'https') === 0;
+            || isset($_SERVER['HTTP_X_FORWARDED_PROTO']) && strncasecmp($_SERVER['HTTP_X_FORWARDED_PROTO'], 'https', 5) === 0;
```
Comment (bpicolo) [inline framework/web/Request.php]: this is magical, happy to s/5/strlen('https') if preferred

**Review by samdark — APPROVED** [review]

**Review by SilverFire — APPROVED** [review]

File: framework/web/Request.php (hunk 1 above)
Comment (samdark) [inline framework/web/Request.php]: It looks OK to me as is.

**Review by dynasource — CHANGES_REQUESTED** [review]
This PR intoduces new unit tests which are not covering all the state possibilities (like for ```_SERVER['HTTPS'], 'on'```). These should be added or at least be commented as 'todo' for future expansion.

Also, it is unclear why strcasecmp is only changed for the second and not for the first use of strcasecmp.

**Comment by bpicolo** [comment]
@dynasource did you read the Issue? The first strcasecmp is irrelevant to the issue at hand.

As for testing, I don't agree that exhaustive testing is or should be a requirement. Especially as test coverage prior is actual 0.

**Comment by dynasource** [comment]
@bpicolo, I am watching your PR from a future code review / read perspective. When someone sees these similar functions a false assumption can be made that a synchronization is forgotten. It would be helpful to understand this from the code with proper naming.

Similar applies for your tests. When somebody is looking at the tests in the future, a false assumption is made about the tests being complete. In this case, it isn't. Also here it would be helpful to easily whats covered and what not. 

We dont want us to burn time in the future to 'interpretate' while this can be prevented.

**Comment by bpicolo** [comment]
I'll change it up, but imo the bar for making OSS contributions convenient here is not being applied at the right height

**Comment by bpicolo** [comment]
At least partly because this was 3 weeks ago and is now much less on mind 😶

**Comment by bpicolo** [comment]
@dynasource  Does that suit you better? I suspect a comment there is more appropriate than revising the former to strncasecmp, and would prefer to avoid temporary variables for naming's sake (yeah, unfortunate that PHP internals naming is poor in general ☹️  )

**Comment by dynasource** [comment]
@bpicolo, thanks for putting effort in the extra work. It will suit if the comment is 1 line and explationary for the use of strncasecmp (optionally by a reference to the issue topic)

**Comment by bpicolo** [comment]
@dynasource  Already put a 4-line comment in to be extra explicit. Seem good?

**Comment by dynasource** [comment]
if you think you cannot shrink this comment any further without affecting interpretation, it is OK by me.

**Comment by bpicolo** [comment]
Wasn't sure if the former was a question or a statement. there ya go

**Review by dynasource — APPROVED** [review]

**Comment by dynasource** [comment]
your persistence is appreciated

**Comment by bpicolo** [comment]
<3 Thanks for the CR. I understand where you're coming from, you know how it is coming back to popups from old contexts at 7AM though, hah.

**Review by samdark — CHANGES_REQUESTED** [review]
Reviewed again. It's not safe to rely on `HTTP_` headers if you're not behind a trusted proxy. I think we should introduce a list of trusted IPs and consider `HTTP_` headers only if requested one is one of these.

**Comment by samdark** [comment]
See https://github.com/yiisoft/yii2/issues/10854

**Comment by bpicolo** [comment]
@samdark seems resolveable by LB configuration no? 

IP whitelist doesn't work well in the world of auto scaling cloud groups.

**Comment by bpicolo** [comment]
True that it's not necessarily a good default though. Not sure.

**Comment by samdark** [comment]
@bpicolo yes, it's resolvable by proper configuration when you have a load balancer in place but isn't a safe default in case you don't have load balancer at all i.e. more than half of all cases.

**Comment by samdark** [comment]
Thanks for contributing. The issue will be solved but not with this pull request. Sorry about that.

**Pull request closed without merging**